File: backend/compiler/compiler.py
import os
import json
import time
from typing import Dict, Any, List

from .intent_extractor import IntentExtractor
from .system_designer import SystemDesigner
from .schema_generator import SchemaGenerator
from .refiner import Refiner
from .repair_engine import RepairEngine
from .execution_simulator import ExecutionSimulator
from .artifact_generator import ArtifactGenerator

class AppCompiler:
    def __init__(self):
        self.intent_extractor = IntentExtractor()
        self.system_designer = SystemDesigner()
        self.schema_generator = SchemaGenerator()
        self.refiner = Refiner()
        self.repair_engine = RepairEngine()
        self.simulator = ExecutionSimulator()

    def compile(self, prompt: str, introduce_anomalies: bool = False) -> Dict[str, Any]:
        print('\n=========================================')
        print('🏁 STARTING COMPILATION PIPELINE (PYTHON)')
        print(f'Prompt: "{prompt}"')
        print('=========================================')

        start = time.time()

        # Stage 1: Intent Extraction Layer
        print('\n[STAGE 1] Lexing and Intent Extraction...')
        intent_ir = self.intent_extractor.extract(prompt)
        print(f"  ✓ Domain matched: {intent_ir.get('domain')}")
        print(f"  ✓ Roles extracted: [{', '.join(intent_ir.get('roles', []))}]")
        print(f"  ✓ Features extracted: [{', '.join(intent_ir.get('features', []))}]")
        if intent_ir.get('ambiguities'):
            print(f"  ⚠ Ambiguities Detected: \n     - " + "\n     - ".join(intent_ir.get('ambiguities')))

        # Stage 2: System Design Layer
        print('\n[STAGE 2] Converting Intent to Architecture IR...')
        architecture_ir = self.system_designer.design(intent_ir)
        print(f"  ✓ Relationships mapped: {len(architecture_ir.get('relationshipGraph', []))} links")
        print(f"  ✓ Routing routes defined: {len(architecture_ir.get('routingTopology', {}).get('routes', []))} paths")

        # Stage 3: Schema Generation Layer
        print('\n[STAGE 3] Schema Generation (DB, API, UI, Auth, Rules)...')
        raw_config = self.schema_generator.generate(intent_ir, architecture_ir, introduce_anomalies)
        print('  ✓ Initial raw schemas provisioned.')

        # Stage 4: Refinement / Validation Layer
        print('\n[STAGE 4] Executing Static Validation Scans...')
        validation_report = self.refiner.validate(raw_config)
        final_config = raw_config
        repair_logs = []

        if not validation_report.get("valid"):
            print(f"  ✗ Validation errors detected: {len(validation_report.get('errors', []))} issues.")
            
            # Stage 5: Targeted Repair Engine
            print('\n[STAGE 5] Triggering Targeted Repair Engine...')
            repair_result = self.repair_engine.repair(raw_config, validation_report)
            final_config = repair_result.get("ast")
            repair_logs = repair_result.get("logs", [])
            
            print('  ✓ Patches applied. Re-running validation...')
            validation_report = self.refiner.validate(final_config)

        if validation_report.get("valid"):
            print('  ✓ Configuration validated: 100% Consistent and Type-Safe.')
        else:
            print('  ✗ Final validation failed. Compilation aborted.')
            return {"success": False, "errors": validation_report.get("errors")}

        # Stage 6: Execution Plan Generation
        execution_plan = self.generate_execution_plan(final_config)

        # Assemble Final Output Document
        compilation_output = {
            "metadata": {
                "compiler_version": 'v1.0.0-prod-python',
                "compilation_time_ms": int((time.time() - start) * 1000),
                "timestamp": time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())
            },
            "intent_ir": intent_ir,
            "architecture_ir": architecture_ir,
            "database_schema": final_config.get("database"),
            "api_schema": final_config.get("api"),
            "ui_schema": final_config.get("ui"),
            "authentication_rules": final_config.get("authentication"),
            "business_logic_rules": final_config.get("business_logic"),
            "validation_reports": {
                "valid": validation_report.get("valid"),
                "errors": validation_report.get("errors")
            },
            "repair_reports": {
                "total_repairs_needed": len(repair_logs),
                "repair_attempts": repair_logs
            },
            "assumptions": intent_ir.get("assumptions", []),
            "execution_plans": {
                "execution_graph_nodes": execution_plan
            }
        }

        # Save configuration build file to output directory
        current_dir = os.path.dirname(os.path.abspath(__file__))
        output_dir = os.path.join(os.path.dirname(current_dir), '..', 'dist')
        os.makedirs(output_dir, exist_ok=True)
        
        output_path = os.path.join(output_dir, 'app_configuration.json')
        with open(output_path, 'w') as f:
            json.dump(compilation_output, f, indent=2)
            
        print(f"\n✓ Emitter: Compiled output saved to {output_path}")

        # Stage 7: Execution Simulation Check
        simulation_result = self.simulator.simulate(final_config)

        # Generate target code files (SQLite, FastAPI, Streamlit)
        generated_files = ArtifactGenerator.generate(final_config, output_dir)
        print(f"✓ Generated executable target application artifacts: {list(generated_files.keys())}")

        return {
            "success": True,
            "outputPath": output_path,
            "compilationOutput": compilation_output,
            "simulationResult": simulation_result,
            "generatedFiles": generated_files
        }

    def generate_execution_plan(self, config: Dict[str, Any]) -> List[Dict[str, Any]]:
        nodes = []
        step_count = 1

        # Table migrations
        tables = config.get("database", {}).get("tables", {})
        for table in tables.keys():
            nodes.append({
                "id": f"STEP_{step_count}_DB_MIGRATION_{table.upper()}",
                "action": 'CREATE_TABLE',
                "description": f"Create database table: \"{table}\"",
                "depends_on": []
            })
            step_count += 1

        # API Route register
        if config.get("api", {}).get("endpoints"):
            nodes.append({
                "id": f"STEP_{step_count}_API_GATEWAY_SYNC",
                "action": 'REGISTER_ROUTES',
                "description": 'Bind endpoint pathways and request validation schemas',
                "depends_on": [n["id"] for n in nodes]
            })
            step_count += 1

        # UI render deploy
        if config.get("ui", {}).get("components"):
            nodes.append({
                "id": f"STEP_{step_count}_UI_FRONTEND_COMPILE",
                "action": 'DEPLOY_VIEWS',
                "description": 'Compile layouts and bind state action data-sources',
                "depends_on": [f"STEP_{step_count - 2}_API_GATEWAY_SYNC"]
            })
            step_count += 1

        return nodes

if __name__ == "__main__":
    import sys
    compiler = AppCompiler()
    default_prompt = "Build a CRM with login, contacts, dashboard, role-based access, and premium plan with payments. Admins can see analytics."
    
    prompt = sys.argv[1] if len(sys.argv) > 1 else default_prompt
    trigger_repair = "--trigger-repair" in sys.argv or "-r" in sys.argv
    
    compiler.compile(prompt, trigger_repair)

File: compiler.js
const fs = require('fs');
const path = require('path');

const IntentExtractor = require('./compiler/intent_extractor');
const SystemDesigner = require('./compiler/system_designer');
const SchemaGenerator = require('./compiler/schema_generator');
const Refiner = require('./compiler/refiner');
const RepairEngine = require('./compiler/repair_engine');
const ExecutionSimulator = require('./compiler/execution_simulator');

class AppCompiler {
  constructor() {
    this.intentExtractor = new IntentExtractor();
    this.systemDesigner = new SystemDesigner();
    this.schemaGenerator = new SchemaGenerator();
    this.refiner = new Refiner();
    this.repairEngine = new RepairEngine();
    this.simulator = new ExecutionSimulator();
  }

  async compile(prompt, introduceAnomalies = false) {
    console.log('\n=========================================');
    console.log('🏁 STARTING COMPILATION PIPELINE');
    console.log(`Prompt: "${prompt}"`);
    console.log('=========================================');

    const start = Date.now();

    // Stage 1: Intent Extraction Layer
    console.log('\n[STAGE 1] Lexing and Intent Extraction...');
    const intentIR = await this.intentExtractor.extract(prompt);
    console.log(`  ✓ Domain matched: ${intentIR.domain}`);
    console.log(`  ✓ Roles extracted: [${intentIR.roles.join(', ')}]`);
    console.log(`  ✓ Features extracted: [${intentIR.features.join(', ')}]`);
    if (intentIR.ambiguities && intentIR.ambiguities.length > 0) {
      console.log(`  ⚠ Ambiguities Detected: \n     - ${intentIR.ambiguities.join('\n     - ')}`);
    }

    // Stage 2: System Design Layer
    console.log('\n[STAGE 2] Converting Intent to Architecture IR...');
    const architectureIR = await this.systemDesigner.design(intentIR);
    console.log(`  ✓ Relationships mapped: ${architectureIR.relationshipGraph.length} links`);
    console.log(`  ✓ Routing routes defined: ${architectureIR.routingTopology.routes.length} paths`);

    // Stage 3: Schema Generation Layer
    console.log('\n[STAGE 3] Schema Generation (DB, API, UI, Auth, Rules)...');
    const rawConfig = await this.schemaGenerator.generate(intentIR, architectureIR, introduceAnomalies);
    console.log('  ✓ Initial raw schemas provisioned.');

    // Stage 4: Refinement / Validation Layer
    console.log('\n[STAGE 4] Executing Static Validation Scans...');
    let validationReport = this.refiner.validate(rawConfig);
    let finalConfig = rawConfig;
    let repairLogs = [];

    if (!validationReport.valid) {
      console.log(`  ✗ Validation errors detected: ${validationReport.errors.length} issues.`);
      
      // Stage 5: Targeted Repair Engine
      console.log('\n[STAGE 5] Triggering Targeted Repair Engine...');
      const repairResult = await this.repairEngine.repair(rawConfig, validationReport);
      finalConfig = repairResult.ast;
      repairLogs = repairResult.logs;
      
      console.log('  ✓ Patches applied. Re-running validation...');
      validationReport = this.refiner.validate(finalConfig);
    }

    if (validationReport.valid) {
      console.log('  ✓ Configuration validated: 100% Consistent and Type-Safe.');
    } else {
      console.error('  ✗ Final validation failed. Compilation aborted.');
      return { success: false, errors: validationReport.errors };
    }

    // Stage 6: Execution Plan Generation
    const executionPlan = this.generateExecutionPlan(finalConfig);

    // Assembly of Final Output Document
    const compilationOutput = {
      metadata: {
        compiler_version: 'v1.0.0-prod',
        compilation_time_ms: Date.now() - start,
        timestamp: new Date().toISOString()
      },
      intent_ir: intentIR,
      architecture_ir: architectureIR,
      database_schema: finalConfig.database,
      api_schema: finalConfig.api,
      ui_schema: finalConfig.ui,
      authentication_rules: finalConfig.authentication,
      business_logic_rules: finalConfig.business_logic,
      validation_reports: {
        valid: validationReport.valid,
        errors: validationReport.errors
      },
      repair_reports: {
        total_repairs_needed: repairLogs.length,
        repair_attempts: repairLogs
      },
      assumptions: intentIR.assumptions,
      execution_plans: {
        execution_graph_nodes: executionPlan
      }
    };

    // Save configuration build file to output directory
    const outputDir = path.join(__dirname, 'dist');
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir);
    }
    const outputPath = path.join(outputDir, 'app_configuration.json');
    fs.writeFileSync(outputPath, JSON.stringify(compilationOutput, null, 2));
    console.log(`\n✓ Emitter: Compiled output saved to ${outputPath}`);

    // Stage 7: Execution Simulation Check
    const simulationResult = this.simulator.simulate(finalConfig);

    return {
      success: true,
      outputPath,
      compilationOutput,
      simulationResult
    };
  }

  generateExecutionPlan(config) {
    const nodes = [];
    let stepCount = 1;

    // Table migrations
    if (config.database.tables) {
      Object.keys(config.database.tables).forEach(table => {
        nodes.push({
          id: `STEP_${stepCount++}_DB_MIGRATION_${table.toUpperCase()}`,
          action: 'CREATE_TABLE',
          description: `Create database table: "${table}"`,
          depends_on: []
        });
      });
    }

    // API Route register
    if (config.api.endpoints) {
      nodes.push({
        id: `STEP_${stepCount++}_API_GATEWAY_SYNC`,
        action: 'REGISTER_ROUTES',
        description: 'Bind endpoint pathways and request validation schemas',
        depends_on: nodes.map(n => n.id)
      });
    }

    // UI render deploy
    if (config.ui.components) {
      nodes.push({
        id: `STEP_${stepCount++}_UI_FRONTEND_COMPILE`,
        action: 'DEPLOY_VIEWS',
        description: 'Compile layouts and bind state action data-sources',
        depends_on: [`STEP_${stepCount - 2}_API_GATEWAY_SYNC`]
      });
    }

    return nodes;
  }
}

// Support executing directly via node CLI
if (require.main === module) {
  const compiler = new AppCompiler();
  const defaultPrompt = "Build a CRM with login, contacts, dashboard, role-based access, and premium plan with payments. Admins can see analytics.";
  
  const userArgs = process.argv.slice(2);
  const prompt = userArgs[0] || defaultPrompt;
  const triggerRepair = userArgs.includes('--trigger-repair') || userArgs.includes('-r');

  compiler.compile(prompt, triggerRepair).catch(err => {
    console.error('Fatal compilation failure:', err);
    process.exit(1);
  });
}

module.exports = AppCompiler;
